### PR TITLE
[6.2.z] Implemented initial test for content management

### PR DESCRIPTION
**Pull request by abalakh**
Depends on SatelliteQE/nailgun#396
Depends on SatelliteQE/robottelo-ci#581

* Updated capsule deployment helper to use correct capsule repo (our internal for downstream, cdn for cdn)
* Updated rhel repo download policy in capsule deployment helper to 'on_demand' (significant performance inrease, syncing time decreased from 40 min to 15 min for me locally)
* Added helper to form local path to published repository (some of our existing tests can use that)
* Added helper to create a repository (not satellite's repository entity, but a group of rpms with repodata)
* Added helper to fetch specific repository's revision
* Added new dir for pulp/content management related tests
* Implemented the first content management test described in #4499

### CODE REVIEW

File: robottelo/vm.py [hunk 1]
```diff
@@ -249,6 +249,9 @@ def install_katello_agent(self):
             raise VirtualMachineError('Failed to install katello-agent')
         gofer_check = self.run('service goferd status')
         if gofer_check.return_code != 0:
+            self.run('service goferd start')
```
Comment (ldjebran) [inline robottelo/vm.py]: perhaps we should add a function parameter to try to start goferd service with False default value, to not break the current tests and in capsule helper call this function with true,

```python
    def install_katello_agent(self, start_goferd=False):
        ...
        if gofer_check.return_code != 0 and start_goferd:
            self.run('service goferd start')
        ...


```

Comment (abalakh) [inline robottelo/vm.py]: Frankly i'm hoping while this PR gets 2 acks and all the issues resolved, katello-agent BZ will become fixed (as it's in progress right now) and i'll just remove these lines 😄 
Otherwise i'd like to insist on adding a workaround here in any acceptable way, as it blocks too much tests for us, even those which aren't directly related (like capsule or content management tests).

**Comment by abalakh** [comment]
Just for the record, to remove `[WIP]` tag, 3 issues have to be resolved:
- `pulp_celerybeat.service` is refusing to start automatically for me. It's most likely env issue as it's working ok for @ldjebran , still, i have to make sure (and remove corresponding workaround when it's resolved)
- katello-agent is not starting automatically and we've got NACK on workaround for it. If we don't change the approach - this PR will be blocked while BZ1431747 is not fixed yet
- for some reason for me repo.sync() doesn't fetch a new repo (after adding it with repocreate) from the first attempt, only from second sync. Need to investigate more on it and either disregard, or fill a bz and add a workaround.

File: robottelo/helpers.py [hunk 2]
```diff
@@ -413,3 +414,86 @@ def default_url_on_new_port(oldport, newport):
...
+    :param str repo_name: repository name - name of a directory with packages
+        and repodata
+    :param str optional hostname: hostname or IP address of the remote host. If
+        ``None`` the hostname will be get from ``main.server.hostname`` config.
+    :return: URL where the repository can be accessed
+    :rtype: str
+    """
+    repo_path = '/var/lib/pulp/published/yum/http/repos'
+    result = ssh.command(
+        'mkdir -p {}/{}'.format(repo_path, repo_name), hostname=hostname)
```
Comment (ldjebran) [inline robottelo/helpers.py]: change naming:
```python
# this can be moved to constants as PULP_PUBLISHED_REPOS_PATH
repos_path = '/var/lib/pulp/published/yum/http/repos'
repo_path = os.path.join(repos_path, repo_name)
result = ssh.command(
        'mkdir -p {}}'.format(repo_path), hostname=hostname)

```

File: robottelo/helpers.py [hunk 3]
```diff
@@ -413,3 +414,86 @@ def default_url_on_new_port(oldport, newport):
...
+    if result.return_code != 0:
+        raise CLIReturnCodeError(
+            result.return_code, result.stderr, 'Unable to create repo dir')
+    for package in packages:
+        result = ssh.command(
+            '(cd {}/{} && wget {}{})'.format(
+                repo_path, repo_name, repo_fetch_url, package),
```
Comment (ldjebran) [inline robottelo/helpers.py]: same as above use os.path.join
and use urljoin for repo_fetch_url and  package

File: robottelo/cli/factory.py [hunk 4]
```diff
@@ -2679,14 +2681,30 @@ def setup_capsule_virtual_machine(capsule_vm, org_id=None, lce_id=None,
         except CLIReturnCodeError as err:
             raise CLIFactoryError(
                 u'Failed to fetch repository info\n{0}'.format(err.msg))
-    # Synchronize the RH repositories
-    for rh_repo in rh_repos:
+    # If we aren't working with CDN, create custom repo with latest capsule
+    # repo available
+    if not settings.cdn and settings.capsule_repo:
```
Comment (ldjebran) [inline robottelo/cli/factory.py]: think we should raise exception when not settings.cdn and not settings.capsule_repo

Comment (abalakh) [inline robottelo/cli/factory.py]: i thought about this, it won't be user-friendly - i suspect some of our folks don't have `capsule_repo` and `sattools_repo` settings set, by introducing exception we'll block them till they update properties file. And also for robottelo users outside our team this change will basically mean 'if you don't want to run the tests against cdn, you have to have downstream repos for satellite tools and satellite capsule' which is probably not the best approach.
Basically, these lines mean 'if we're in downstream then try to use internal downstream build if available, if not - at least use latest released one' (which will work in most cases btw).

Comment (ldjebran) [inline robottelo/cli/factory.py]: @abalakh , so by this logic you will end without exception and without the necessary repos in content view

Comment (abalakh) [inline robottelo/cli/factory.py]: @ldjebran good catch, i've messed with commits. The checks above should be `if setting.cdn or not settings.capsule_repo`. This way repos will be included.

Comment (abalakh) [inline robottelo/cli/factory.py]: @ldjebran i've updated #4555 as that's a better place to introduce this change. Let's continue our discussion there.

File: robottelo/helpers.py (hunk 3 above)
Comment (abalakh) [inline robottelo/helpers.py]: No objections on using os.path.join, but what's the benefit of using urljoin in this particular case? I'm not constructing url from different parts here, i'm building a new one using scheme, hostname, port and static path. Can't see any use in urljoin here, maybe you meant urlunparse or urlunsplit (which are overkill here too, imo)?

Comment (abalakh) [inline robottelo/helpers.py]: Just curious - what's the benefit of using urlparse here? It has the same drawback:
```python
In [20]: urljoin('http://example.com/some/repo', 'package.rpm')
Out[20]: 'http://example.com/some/package.rpm'
```
I'll add extra check on trailing slash, but so far switching to urljoin makes no sense for me

**Comment by abalakh** [comment]
Depends on #4555

**Comment by abalakh** [comment]
Recent changes:
* Removed invalid assertions (revision id on satellite will always change after repo sync)
* Fixed bug with wrong owner for repo folder on satellite
* Added workaround for BZ1446930
* Updated test to be able to re-use existing capsule
* Moved capsule creation/deletion to setUpClass/tearDownClass accordingly

**Comment by abalakh** [comment]
Made a squash of all the commits, rebase and basically PR is ready for review. Test results with reusing existing capsule:
```python
py.test -v tests/foreman/api/test_contentmanagement.py
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/andrii/workspace/env62z/bin/python
cachedir: .cache
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.3.1
collected 1 items
2017-05-05 14:33:07 - conftest - DEBUG - Found WONTFIX in decorated tests ['1110476', '1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-05-05 14:33:07 - conftest - DEBUG - Collected 1 test cases


Implemented initial test for content management


tests/foreman/api/test_contentmanagement.py::ContentManagementTestCase::test_positive_capsule_sync PASSED

============================== 0 tests deselected ==============================
========================== 1 passed in 336.04 seconds ==========================
```
Will post test results without reusing capsule in ~1h.

**Comment by abalakh** [comment]
Test results:
```python
py.test -v tests/foreman/api/test_contentmanagement.py
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/andrii/workspace/env/bin/python
cachedir: .cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.3.1
collected 1 items
2017-05-05 14:57:45 - conftest - DEBUG - Found WONTFIX in decorated tests ['1110476', '1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-05-05 14:57:45 - conftest - DEBUG - Collected 1 test cases


tests/foreman/api/test_contentmanagement.py::ContentManagementTestCase::test_positive_capsule_sync PASSED

============================== 0 tests deselected ==============================
========================= 1 passed in 2172.58 seconds ==========================
```

File: robottelo/helpers.py [hunk 5]
```diff
@@ -448,3 +451,89 @@ def get_services_status():
...
+    :param str repo_fetch_url: URL to fetch packages from
+    :param packages: list of packages to fetch (with extension)
+    :param str repo_name: repository name - name of a directory with packages
+        and repodata
+    :param str optional hostname: hostname or IP address of the remote host. If
+        ``None`` the hostname will be get from ``main.server.hostname`` config.
+    :return: URL where the repository can be accessed
+    :rtype: str
+    """
+    base_path = '/var/lib/pulp/published/yum/http/repos'
```
Comment (ldjebran) [inline robottelo/helpers.py]: think this can be a subject of a constant variable
```python
PULP_PUBLISHED_YUM_REPOS_PATH = '/var/lib/pulp/published/yum/http/repos'
```
what do you think?
